detectors/node/opentelemetry-resource-detector-alibaba-cloud/src/detectors/AlibabaCloudEcsDetector.ts

```diff
@@ -0,0 +1,142 @@
+  readonly MILLISECOND_TIME_OUT = 1000;
+
+  /**
+   * Attempts to connect and obtain an AlibabaCloud instance Identity document.
+   * If the connection is successful it returns a promise containing a
+   * {@link Resource} populated with instance metadata. Returns a promise
+   * containing an empty {@link Resource} if the connection or parsing of the
+   * identity document fails.
+   *
+   * @param config (unused) The resource detection config
+   */
+  async detect(_config?: ResourceDetectionConfig): Promise<Resource> {
+    const {
+      'owner-account-id': accountId,
+      'instance-id': instanceId,
+      'instance-type': instanceType,
+      'region-id': region,
+      'zone-id': availabilityZone,
+    } = await this._fetchIdentity();
+    const hostname = await this._fetchHost();
+
+    return new Resource({
+      [SemanticResourceAttributes.CLOUD_PROVIDER]:
+        CloudProviderValues.ALIBABA_CLOUD,
+      [SemanticResourceAttributes.CLOUD_PLATFORM]:
+        CloudPlatformValues.ALIBABA_CLOUD_ECS,
+      [SemanticResourceAttributes.CLOUD_ACCOUNT_ID]: accountId,
+      [SemanticResourceAttributes.CLOUD_REGION]: region,
+      [SemanticResourceAttributes.CLOUD_AVAILABILITY_ZONE]: availabilityZone,
+      [SemanticResourceAttributes.HOST_ID]: instanceId,
+      [SemanticResourceAttributes.HOST_TYPE]: instanceType,
+      [SemanticResourceAttributes.HOST_NAME]: hostname,
+    });
+  }
+
+  /**
+   * Fetch AlibabaCloud instance document url with http requests. If the
+   * application is running on an ECS instance, we should be able to get back a
+   * valid JSON document. Parses that document and stores the identity
+   * properties in a local map.
+   */
+  private async _fetchIdentity(): Promise<any> {
+    const options = {
+      host: this.ALIBABA_CLOUD_IDMS_ENDPOINT,
+      path: this.ALIBABA_CLOUD_INSTANCE_IDENTITY_DOCUMENT_PATH,
+      method: 'GET',
+      timeout: this.MILLISECOND_TIME_OUT,
+    };
+    const identity = await this._fetchString(options);
+    return JSON.parse(identity);
+  }
+
+  private async _fetchHost(): Promise<string> {
+    const options = {
+      host: this.ALIBABA_CLOUD_IDMS_ENDPOINT,
+      path: this.ALIBABA_CLOUD_INSTANCE_HOST_DOCUMENT_PATH,
+      method: 'GET',
+      timeout: this.MILLISECOND_TIME_OUT,
+    };
+    return await this._fetchString(options);
+  }
+
+  private async _fetchString(options: http.RequestOptions): Promise<string> {
+    return new Promise((resolve, reject) => {
+      const timeoutId = setTimeout(() => {
+        req.abort();
+        reject(new Error('ECS metadata api request timed out.'));
+      }, 1000);
+
+      const req = http.request(options, res => {
+        clearTimeout(timeoutId);
+        const { statusCode } = res;
+        res.setEncoding('utf8');
+        let rawData = '';
+        res.on('data', chunk => (rawData += chunk));
+        res.on('end', () => {
+          if (statusCode && statusCode >= 200 && statusCode < 300) {
```

Comment:
You could check the status code earlier and discard the incoming data (by calling `res.resum()`) in that case.